gnomAD updates

mwalker174 opened the pull request:
A number of bug fixes encountered while processing gnomAD-v3, 1KGP, and GMKF. This includes the replacement of tabix with GATK for streaming evidence records from buckets (with suggested changes incorporated from #26). Docker updates forthcoming. Addresses #28.

- Add GATK to sv-base docker; use GATK for PE/SR/RD localization
- fix fragile cleanvcf1 block
- Fix BedCluster too
- Use CHR2 for end locus in svtk count-sr
- Pad RDTest bincov localization; fix inputs for cromwell 51
- Correct bounds for RDTestChromosome
- Update to PrintSVEvidence and start fixing WDLs for cromwell v51
- Fix sv-base dockerfile
- Fix data.table dependency
- Switch to .RD.txt.gz extension for depth evidence files
- Fix localized filenames in 02 TestChromosome tasks
- Fix tabix arguments and a few evidence file names
- random forest fix: select cutoff from AUC and FDR
- fixes to SR_genotype.opt_part2.sh
- bug fixed: 1. runtime_override_collect_vids_per_sample were not passed in Module05_06.wdl; 2. wrong white list were used in RdGenotype
- unnecessary functions removed
- -a added for ConcatVcf
- Module07-09 wdl scripts renamed
- Prevent empty PE and SR common site test shards in 02

cwhelan commented on wdl/PETestChromosome.wdl:
```diff
@@ -101,36 +101,54 @@ workflow PETestChromosome {
+    }
+  }
+
+  if (!allosome) {
```
Some of these changes might conflict with my proposed changes to support empty shards in https://github.com/broadinstitute/gatk-sv/pull/54 -- just an fyi. It'd be nice if we can make sure the resulting code after merges doesn't error on empty shards.

mwalker174 replied:
Yes I'll do a rebase of this branch after yours is merged and make sure the empty shard check is in place

cwhelan reviewed (commented):
Here are some comments; I held off on looking at the contents of module07 and 08 since it sounds like that might still be in flux. It'd be good to rename the current `module07.wdl` which is annotations, to `module08.wdl`, maybe in a separate commit?

mwalker174 commented on src/sv-pipeline/04_variant_resolution/scripts/clean_vcf_part1.sh:
```diff
@@ -27,18 +27,27 @@ zcat $vcf \
+
+echo '##FORMAT=<ID=EV,Number=1,Type=String,Description="Classes of evidence supporting final genotype">' > ev_header.txt
+tabix $vcf
+bcftools annotate -x FORMAT/EV $vcf | bgzip -c > no_ev.vcf.gz
```
It turns out that wasn't possible because the `Type` is getting changed from Integer to String.

mwalker174 commented on src/svtk/svtk/adjudicate/random_forest.py:
```diff
@@ -120,10 +120,16 @@ def learn_cutoffs(self):
         cutoff_metrics['pass_cutoffs'] = True
         passing = cutoff_metrics['pass_cutoffs']
 
+        print("evidence: " +self.name)
```
Must have been for debugging, so I'll delete

mwalker174 commented on src/svtk/svtk/adjudicate/random_forest.py:
```diff
@@ -178,3 +189,27 @@ def learn_cutoff(metric, probs):
 
     return thresh[best_idx]
+
+def learn_cutoff_V2(metric, probs,fdr_cff=.05):
```
Changed `learn_cutoff` -> `learn_cutoff_dist` and `learn_cutoff_V2` -> `learn_cutoff_fdr`

mwalker174 commented on src/svtk/svtk/cli/pesr_test.py:
```diff
@@ -309,11 +309,11 @@ def count_sr(argv):
+                pos, strand, chrom = record.pos, record.info['STRANDS'][0], record.chrom
             else:
-                pos, strand = record.stop, record.info['STRANDS'][1]
+                pos, strand, chrom = record.stop, record.info['STRANDS'][1], record.info['CHR2']
```
Done

mwalker174 commented on wdl/AnnotateVcf.wdl:
```diff
@@ -97,36 +97,52 @@ task SubsetVcf {
+      -L "~{contig}" \
+      -O ~{prefix}.~{contig}.vcf
+
+    # GATK does not block compress
```
Thanks for pointing this out. It depends on the way the tool is implemented, but I just tested SelectVariants and it does seem to support block compressed output and actually generates an index automatically, so this block of code isn't needed.

mwalker174 commented on wdl/Module07FilterCleanupQualRecalibration.wdl:
```diff
@@ -0,0 +1,242 @@
+  RuntimeAttr runtime_attr = select_first([runtime_attr_override, default_attr])
+  command <<<
+    set -euo pipefail
+    #zcat ~{mcnv} |uniq | bgzip > mcnv.vcf.gz
```
Done

mwalker174 commented on wdl/Module07FilterCleanupQualRecalibration.wdl:
```diff
@@ -0,0 +1,242 @@
+      cp all.samples.list pcrminus.samples.list
+    fi
+    #Restrict famfiles
+    #while read ptn; do fgrep -w $ptn ~{famfile}; done < all.samples.list > revised.fam
```
Done

mwalker174 commented:
Thanks for the review @cwhelan. Somehow there were a number of duplicated WDLs for modules 07-09, so I cleaned those up. Also updated test jsons for former 07 to be 08. 

I addressed most of your comments, but I still want to hold off on modifying 07 and 09 since I believe they are working as is, and we don't have jsons for testing changes right now. I opened issue #55 to track that we need to review these modules still.

cwhelan commented on wdl/Module07XfBatchEffect.wdl:
```diff
@@ -0,0 +1,679 @@
+    done < <( sed -n '1p' input.list ) \
+    > header.txt
+    while read batch file; do
+      for wrapper in 1; do
```
This for loop seems unnecessary to me.

cwhelan commented on wdl/Module07XfBatchEffect.wdl:
```diff
@@ -0,0 +1,679 @@
+        | sort -Vk1,1 \
+        | cut -f1-3
+    done < <( sed -n '1p' input.list ) \
+    > header.txt
```
This doesn't seem like just a header, it gets pasted to the processed files.. I guess I think of header as always being column headers, maybe calling this row_headers.txt would make this more clear.

cwhelan commented on wdl/Module07XfBatchEffect.wdl:
```diff
@@ -0,0 +1,679 @@
+    max_retries: 1
+  }
+  RuntimeAttr runtime_attr = select_first([runtime_attr_override, default_attr])
+  command <<<
```
If it works now that's fine, but I feel like this task could be greatly simplified by either using the `join` command or a small python script.

cwhelan commented on wdl/Module07XfBatchEffect.wdl:
```diff
@@ -0,0 +1,679 @@
+    #Get master list of PCR- to PCR- failures   #removed from the PCR- only projects
+    #Get master list of PCR+ to PCR- failures   #removed from the PCR- only projects
+    #Get master list of all possible failures
+    cat ~{write_lines(fail_variant_lists)} \
```
Are there so many of these files that you can't use `{sep=" " ...}`?

cwhelan commented on wdl/Tasks04.wdl:
```diff
@@ -394,20 +415,30 @@ task CountPE {
-    awk -v OFS="\t" -v window=5000 '{if ($2-window>0){print $1,$2-window,$2+window}else{print $1,0,$2+window}}' test.bed  >> region.bed
-    awk -v OFS="\t" -v window=5000 '{if ($3-window>0){print $1,$3-window,$3+window}else{print $1,0,$3+window}}' test.bed  >> region.bed
+    awk -v OFS="\t" -v window=500 '{if ($2-window>0){print $1,$2-window,$2+window}else{print $1,0,$2+window}}' test.bed  > region.bed
+    awk -v OFS="\t" -v window=500 '{if ($3-window>0){print $1,$3-window,$3+window}else{print $1,0,$3+window}}' test.bed  >> region.bed
```
Is 500 bases of padding always sufficient to get the evidence we needed? 

Should we parameterize this value? I could also see eventually using a value computed from the data (based on insert size metrics distributions, for example).

mwalker174 replied:
500 is default the "outer" padding for `svtk count-pe` so hard-coding is okay in this case. 

It would be better to adjust this per-batch based on insert size distribution, but this would take a bit of work to optimize. It's not obvious how to do that given that we don't usually batch based on the mean or variance of the insert size distribution, although they may correlate with dosage score in some way.

cwhelan commented on dockerfiles/sv-base/Dockerfile:
```diff
@@ -7,6 +7,33 @@ FROM ${MINIBASE_IMAGE}
 ENV HOME=/root
 
+####################################################################################################
+# GATK needed for PrintSVEvidence
```
It might be helpful to make a ticket to update to a gatk release rather than this commit once PrintSVEvidence is merged into GATK master.

mwalker174 replied:
Created #81

cwhelan reviewed (approved):
This looks good to me in general. I know you've tested this out on real data so take my comments more as suggestions than requirements to fix. There are some places in the bash tasks that look overly complicated to my eye but perhaps you ran into issues that required the complexity -- it might be good to sprinkle in some comments about why things are implemented the way they are. As noted earlier, I think there are some places where bcftools would do things more efficiently and enforce valid VCF -- perhaps you could expand that ticket to note some of them.

mwalker174 commented on README.md:
```diff
@@ -73,6 +73,12 @@ Example workflow inputs can be found in `/inputs`. All required resources are av
+**Important**: The following parameters must be set when certain input data is in requester pays (RP) buckets:
+
+* `GATKSVPipelineSingleSample.requester_pays_cram` and `GATKSVPipelineBatch.Module00aBatch.requester_pays_crams` - set to `True` if inputs are CRAM format and in an RP bucket, otherwise `False`.
+* `GATKSVPipelineBatch.GATKSVPipelinePhase1.gcs_project_for_requester_pays` - set to your Google Cloud Project ID if gVCFs are in an RP bucket, otherwise omit this parameter.
```
samtools doesn't support RP so `CramToBamRequesterPays` doesn't localize the CRAM.

cwhelan replied:
Oh, of course, sorry..

mwalker174 commented on inputs/GATKSVPipelineSingleSample.ref_panel_1kg.na12878.json:
```diff
@@ -4,40 +4,40 @@
   "GATKSVPipelineSingleSample.requester_pays_cram": "true",
 
-  "GATKSVPipelineSingleSample.cutoffs" : "gs://gcp-public-data--broad-references/hg38/v0/sv-resources/ref-panel/1KG/v1/cutoffs/ref_panel_1kg_v1.cutoffs",
+  "GATKSVPipelineSingleSample.cutoffs" : "gs://broad-dsde-methods-markw/gatksv/mw-gnomad-v2/1kg_batch_outputs/ref_panel_1kg_v1.cutoffs",
```
Oops, I had forgotten to move these. I moved all newly generated files to `gs://gatk-sv-resources-public/hg38/v0/sv-resources/ref-panel/1KG/v2/` and renamed the reference panel from `ref_panel_1kg_v1` to `ref_panel_1kg_v2`. Anything currently in `gs://gcp-public-data--broad-references` I am keeping the same to reduce transfer cost burden on the Methods project, even though these are labeled as `v1`.

mwalker174 commented on inputs/GATKSVPipelineSingleSample.ref_panel_1kg.na12878.json:
```diff
@@ -259,1080 +259,1080 @@
-    "gs://gcp-public-data--broad-references/hg38/v0/sv-resources/ref-panel/1KG/v1/manta/std_153.manta.NA21102.vcf.gz",
-    "gs://gcp-public-data--broad-references/hg38/v0/sv-resources/ref-panel/1KG/v1/manta/std_154.manta.NA21122.vcf.gz",
-    "gs://gcp-public-data--broad-references/hg38/v0/sv-resources/ref-panel/1KG/v1/manta/std_155.manta.NA21133.vcf.gz"
+    "gs://broad-dsde-methods-markw/gatksv/mw-gnomad-v2/1kg_batch_outputs/std_000.manta.HG00096.vcf.gz",
```
Done

mwalker174 commented on wdl/DepthClustering.wdl:
```diff
@@ -227,7 +227,11 @@ task BedCluster {
+      "bedtools coverage -a ~{batch}.~{svtype}.~{chrom}.preexcludelist.bed -b ~{exclude_list} | awk '$NF < ~{exclude_list_frac_max}' | rev | cut -f5- | rev > excluded.filtered.bed"
+      else
+      ""}
+    ~{if defined(exclude_list) then
```
Yes, as far as I know you are only allowed inline `if` statements inside of the command block.

mwalker174 commented on wdl/Module07FilterCleanupQualRecalibration.wdl:
```diff
@@ -0,0 +1,241 @@
+## EXPERIMENTAL WORKFLOW
+##########################
+
+# Based on https://portal.firecloud.org/#methods/Talkowski-SV/filter_cleanup_and_QUAL_recalibration_CMC_copy/7
```
It's probably because you don't have read permission (I get an error as well).

mwalker174 commented on wdl/Module07FilterOutlierSamples.wdl:
```diff
@@ -0,0 +1,410 @@
+
+  # Get outliers
+  if (PCRPLUS) {
+    call IdentifyOutliers as identify_PCRPLUS_outliers {
```
Done

mwalker174 commented on wdl/Module07FilterOutlierSamples.wdl:
```diff
@@ -0,0 +1,410 @@
+  RuntimeAttr runtime_attr = select_first([runtime_attr_override, default_attr])
+  command <<<
+    set -euo pipefail
+    tabix --print-header "~{vcf}" "~{contig}" \
```
Let's roll this into #78, I've updated the title.

mwalker174 commented:
Thanks @cwhelan! There is quite a bit of work to do on Module07 and so I have renamed #78 more generally. We do need to move ahead with gnomAD ASAP so I am going to punt this down the road and merge.
